src/lib.rs

```diff
@@ -107,3 +118,46 @@ unsafe fn secbuf_desc(bufs: &mut [sspi::SecBuffer]) -> sspi::SecBufferDesc {
         pBuffers: bufs.as_mut_ptr(),
     }
 }
+
+fn alpn_list(protos: &[Vec<u8>]) -> Vec<u8> {
+    // ALPN wire format is each ALPN preceded by its length as a byte.
+    let mut alpn_wire_format = Vec::with_capacity(
+        protos.iter().map(Vec::len).sum::<usize>() + protos.len(),
+    );
+    for alpn in protos {
+        alpn_wire_format.push(alpn.len() as u8);
+        alpn_wire_format.extend(alpn);
+    }
+
+    // Make sure that the memory we're using for `sspi::SEC_APPLICATION_PROTOCOLS` matches
+    // its current 4-byte alignment requirements.
+    let size = SEC_APPLICATION_PROTOCOL_HEADER_SIZE + alpn_wire_format.len();
+    let mut aligned = Vec::<u32>::with_capacity((size - 1) / mem::size_of::<u32>() + 1);
+    let p = aligned.as_mut_ptr() as *mut u8;
+    let cap = aligned.capacity() * (mem::size_of::<u32>() / mem::size_of::<u8>());
+
+    mem::forget(aligned);
+
+    unsafe {
+        let mut buf = Vec::from_raw_parts(p, 0, cap);
```

Comment:
This is undefined behavior and depending on the used allocator might fail.
`ptr's T needs to have the same size and alignment as it was allocated with.`
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts

Reply:
That's a really good catch, I hadn't realized this before. I'll make a new `struct AlpnList` and use the allocator API.